### PR TITLE
Send transaction building

### PR DESCRIPTION
**Pull request by argjv**
BG-15589 Add support for Tezos multisig wallet transfers.

Changes include:
- Add multisig send transaction support
- Add singlesig send transaction support
- Change `Entry` `value` field name to `amount` and type to `string`
- Move Tezos core signing method to Utils
- Split validator methods for implicit and originated accounts

### CODE REVIEW

File: src/coin/xtz/transactionBuilder.ts [hunk 1]
```diff
@@ -101,15 +113,45 @@ export class TransactionBuilder extends BaseTransactionBuilder {
...
+            transactionOp = transactionOperation(
+              this._counter.toString(),
+              this._sourceAddress,
+              transfer.fee.fee,
+              transfer.fee.gasLimit || DEFAULT_GAS_LIMIT.TRANSFER.toString(),
+              transfer.fee.storageLimit || DEFAULT_STORAGE_LIMIT.TRANSFER.toString(),
+              transfer.amount,
+              transfer.to,
+            );
+          }
```
Comment (CapnMigraine) [inline src/coin/xtz/transactionBuilder.ts]: nit: avoid duplicating lines 129-136 with something like:

```
        this._transfers.forEach(async t => {
          const transfer = t.build();
          let signatures;
          if (transfer.dataToSign) {
            signatures = await this.getSignatures(transfer.dataToSign);
          }
          const transactionOp = transactionOperation(
            this._counter.toString(),
            this._sourceAddress,
            transfer.fee.fee,
            transfer.fee.gasLimit || DEFAULT_GAS_LIMIT.TRANSFER.toString(),
            transfer.fee.storageLimit || DEFAULT_STORAGE_LIMIT.TRANSFER.toString(),
            transfer.amount,
            transfer.to,
            transfer.from,
            signatures,
          );

File: package.json [hunk 2]
```diff
@@ -18,8 +18,8 @@
     "gen-coverage": "./node_modules/.bin/nyc report --reporter=text-lcov > coverage.lcov",
     "gen-protobuf": "node node_modules/protobufjs/cli/bin/pbjs -t static-module -w commonjs -o ./resources/trx/protobuf/tron.js ./resources/trx/protobuf/Discover.proto ./resources/trx/protobuf/Contract.proto ./resources/trx/protobuf/tron.proto",
     "gen-protobufts": "node node_modules/protobufjs/cli/bin/pbts -o ./resources/trx/protobuf/tron.d.ts ./resources/trx/protobuf/tron.js",
-    "lint": "eslint 'src/**/*.ts' && eslint 'test/**/*.ts' || true",
-    "lint-fix": "eslint 'src/**/*.ts' --fix && eslint 'test/**/*.ts' --fix || true",
+    "lint": "eslint 'src/**/*.ts' && eslint 'test/**/*.ts' && eslint 'resources/xtz/**/*.ts' || true",
+    "lint-fix": "eslint 'src/**/*.ts' --fix && eslint 'test/**/*.ts' --fix  && eslint 'resources/xtz/**/*.ts' --fix || true",
```
[the remark this answers is not in the document]
Comment (argjv) [reply package.json]: I don't want to modify the automated code for TRX, but I do want to format the XTZ one.

File: resources/xtz/multisig.ts [hunk 3]
```diff
@@ -1,4 +1,279 @@
...
+import { Entry } from '../../src/coin/baseCoin/iface';
+
+/**
+ * The the entry values from a transaction operation on a generic multisig smart contract.
+ *
+ * @param {TransactionOp} operation A transaction operation JSON
+ * @return {Entry} Information about the destination, token and transfer amount
+ */
+export function getTransferDataFromOperation(operation: TransactionOp): Entry {
+  const transferArgs = operation.parameters.value.args[0].args[1].args[0];
```
[the remark this answers is not in the document]
Comment (argjv) [reply resources/xtz/multisig.ts]: This is from: genericMultisigTransferParams()

File: resources/xtz/multisig.ts [hunk 4]
```diff
@@ -1,4 +1,279 @@
...
+export function transactionOperation(
+  counter: string,
+  source: string,
+  fee: string,
+  gasLimit: string,
+  storageLimit: string,
+  amount: string,
+  destination: string,
+  contractAddress?: string,
+  signatures: string[] = [],
```
[the remark this answers is not in the document]
Comment (argjv) [reply resources/xtz/multisig.ts]: If it has a default value, it is implicitly optional

File: resources/xtz/multisig.ts [hunk 5]
```diff
@@ -1,4 +1,279 @@
...
+  const transactionSignatures: any[] = [];
+  signatures.forEach(s => transactionSignatures.push({ string: s }));
+  return {
+    entrypoint: 'main',
+    value: {
+      prim: 'Pair',
+      args: [
+        {
+          prim: 'Pair',
+          args: [{ int: '0' }, { prim: 'Left', args: [transferToImplicitAccount(destinationAddress, amount)] }],
```
[the remark this answers is not in the document]
Comment (argjv) [reply resources/xtz/multisig.ts]: Implicit or originated. Fixed the methods names.

File: src/coin/xtz/utils.ts [hunk 6]
```diff
@@ -58,6 +62,41 @@ export async function calculateOriginatedAddress(transactionId: string, index: n
...
+
+/**
+ * Useful wrapper to create the generic multisig contract data to sign when moving funds.
+ *
+ * @param {string} contractAddress The wallet contract address with the funds to withdraw
+ * @param {string} destinationAddress The address to transfer the funds to
+ * @param {number} amount Number mutez to transfer
+ * @return {any} A JSON representation of the Michelson script to sign and approve a transfer
+ */
+export function generateDataToSign(contractAddress: string, destinationAddress: string, amount: string): any {
```
[the remark this answers is not in the document]
Comment (argjv) [reply src/coin/xtz/utils.ts]: This is exposed in Tezos Utils so the user can pack the data to be signed. Taquito does not have a method to build the packedData needed.

File: src/coin/xtz/transactionBuilder.ts (hunk 1 above)
Comment (argjv) [reply src/coin/xtz/transactionBuilder.ts]: I'm not sure this will work, `transfer.from` is always present

File: src/coin/trx/transaction.ts [hunk 7]
```diff
@@ -54,25 +54,25 @@ export class Transaction extends BaseTransaction {
     switch (rawData.contractType) {
       case ContractType.Transfer:
         this._type = TransactionType.Send;
-        const value = new BigNumber((rawData.contract[0] as TransferContract).parameter.value.amount);
+        const amount = new BigNumber((rawData.contract[0] as TransferContract).parameter.value.amount).toFixed(0);
         output = {
           address: (rawData.contract[0] as TransferContract).parameter.value.to_address,
-          value,
+          amount,
```
Comment (twtaylorbitgo) [inline src/coin/trx/transaction.ts]: I don't understand why the renaming from value to amount? For clarity?

Comment (argjv) [reply src/coin/trx/transaction.ts]: Yeah, but I reverted it to align with what we use in platform

**Pull request merged**